library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/impl/event/EventPhaseData.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2016, 2017, 2018, 2019 FabricMC
+ * Copyright 2021 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.base.impl.event;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.jetbrains.annotations.ApiStatus;
+
+import net.minecraft.util.Identifier;
+
+/**
+ * Data of an {@link org.quiltmc.qsl.base.api.event.Event} phase.
+ */
+@ApiStatus.Internal
+public final class EventPhaseData<T> {
+	final Identifier id;
+	T[] listeners;
+	final List<EventPhaseData<T>> subsequentPhases = new ArrayList<>();
+	final List<EventPhaseData<T>> previousPhases = new ArrayList<>();
+	int visitStatus = 0; // 0: not visited, 1: visiting, 2: visited
```

Comment:
What's the advantage of an int over an enum here? Is it much faster?

Reply:
Switched to an enum, there should not be any changes to how fast it goes, and it's code that runs only on callback registration.
Enum should be more readable.